types/validator_set.go

```diff
@@ -330,54 +359,45 @@ func (valSet *ValidatorSet) VerifyCommitAny(newSet *ValidatorSet, chainID string
 		if precommit.Type != VoteTypePrecommit {
 			return cmn.NewError("Invalid commit -- not precommit @ index %v", idx)
 		}
-		if !blockID.Equals(precommit.BlockID) {
-			continue // Not an error, but doesn't count
-		}
-
-		// we only grab by address, ignoring unknown validators
-		vi, ov := valSet.GetByAddress(precommit.ValidatorAddress)
-		if ov == nil || seen[vi] {
+		// See if this validator is in oldVals.
+		idx, val := oldVals.GetByAddress(precommit.ValidatorAddress)
+		if val == nil || seen[idx] {
 			continue // missing or double vote...
 		}
-		seen[vi] = true
+		seen[idx] = true
 
-		// Validate signature old school
+		// Validate signature.
 		precommitSignBytes := precommit.SignBytes(chainID)
-		if !ov.PubKey.VerifyBytes(precommitSignBytes, precommit.Signature) {
+		if !val.PubKey.VerifyBytes(precommitSignBytes, precommit.Signature) {
 			return cmn.NewError("Invalid commit -- invalid signature: %v", precommit)
 		}
 		// Good precommit!
-		oldVotingPower += ov.VotingPower
-
-		// check new school
-		_, cv := newSet.GetByIndex(idx)
-		if cv.PubKey.Equals(ov.PubKey) {
-			// make sure this is properly set in the current block as well
-			newVotingPower += cv.VotingPower
+		if blockID.Equals(precommit.BlockID) {
+			oldVotingPower += val.VotingPower
+		} else {
+			// It's OK that the BlockID doesn't match.  We include stray
+			// precommits to measure validator availability.
 		}
 	}
 
-	if oldVotingPower <= valSet.TotalVotingPower()*2/3 {
+	if oldVotingPower <= oldVals.TotalVotingPower()*2/3 {
```

Comment:
I am not sure if this condition is sufficient. Given trusted valsetA at height A, we can trust newValSet at height A+1 as >2/3 of the voting power of our trusted valSet signed the valSet change. From that point, newValSet is trusted valSet and we assume that at most 1/3 of voting power of the new valSet will be faulty. If we consider some valSetB at height B > A+1, it can be quite different than our trusted valSet, for example it can add significant number of new validators and significant new voting power. So the trusted valset can be small portion of valSetB. What security guarantee we have from saying that >2/3 of valSetA signed header for height B as it can be 10 percent of valSetB? Maybe the condition should be that we accept header for B only if validators from our trusted valSet signed >2/3 of valSetB.